lib/Bedrock/ConstantAlreadyDefinedException.php

```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Roots\Bedrock;
+
+/**
+ * Class ConstantAlreadyDefinedException
+ * This should be thrown when a user attempts to define() a constant that has already been defined
+ * @package Roots\Bedrock
+ */
+class ConstantAlreadyDefinedException extends \RuntimeException
```

Comment:
Should we use `LogicException` instead?

Because `LogicException` is
> Exception that represents error in the program logic. This kind of exception should **lead directly to a fix in your code**. 
>
> https://secure.php.net/manual/en/class.logicexception.php

Same goes to `UndefinedConfigKeyException`

Reply:
Hey! Thank you for taking a look at this. 

I understand where you are coming from suggesting when suggesting \LogicException. However, I am still going to advocate for \RuntimeException.

In general when trying to determine what exception class to extend I always look what other exceptions subclass it. Let's look at the exception hierarchy with some cherry-picked examples:

- LogicException (extends Exception): Exception that represents error in the program logic. This kind of exception should lead directly to a fix in your code.
    - BadFunctionCallException: 
    - BadMethodCallException
    - DomainException
    - InvalidArgumentException
    - LengthException
    - OutOfRangeException: Exception thrown when an illegal index was requested. **This represents errors that *should be* detected at compile time**.
- RuntimeException (extends Exception): Exception thrown if **an error which can only be found on runtime occurs**.
    - OutOfBoundsException: Exception thrown if a value is not a valid key. **This represents errors that *cannot be* detected at compile time**.
    - OverflowException
    - RangeException
    - UnderflowException
    - UnexpectedValueException

The fact that the runtime has been polluted with unexpected constants is something that cannot be detected at compile time. Especially since we dynamically build this configuration based on environment variables.

Take a look at the get method of \DS\Map: https://secure.php.net/manual/en/ds-map.get.php. This throws an OutOfBoundsException (extends RuntimeException) if the key could not be found and a default value was not provided.

Lastly consider the following completely valid use case:

JankyHomebrewDBClient.php
```
if (!defined('DB_PORT')) {
  define('DB_PORT', 3306)
}

class JankyHomebrewDBClient {
  // ...
}
```

config/application.php
```
// ...
Config::define('DB_PORT', env('DB_PORT'));
// ...
Config::apply();
```

one-off-db-migration-script.php
```php
// I just want to connect to the DB I don't need WordPress
require_once __DIR__ . '/vendor/autoload.php';
require_once __DIR__ . '/config/application.php';

$client = new JankyHomebrewDBClient(/* ... */);
```

This will throw a `ConstantAlreadyDefinedException` at runtime. I don't necessarily consider this a logical error, since JankyHomebrewDBClient can never tell if it is going to define DB_PORT at compile time.

What do you think?

Reply:
Awesome! This reply should go into textbooks as "best code review reply example".
Thanks!!!